File: core/core-domain/src/main/java/com/coffee/gu/payment/PaymentApprovalResult.java
package com.coffee.gu.payment;

import com.coffee.gu.enums.OrderState;
import com.coffee.gu.enums.PaymentState;

import java.time.OffsetDateTime;

public class PaymentApprovalResult {
    private String orderKey;
    private String paymentKey;
    private PaymentState paymentState;
    private OrderState orderState;
    private Boolean idempotent;
    private OffsetDateTime approvedAt;

    public PaymentApprovalResult(String orderKey, String paymentKey, PaymentState paymentState, OrderState orderState, Boolean idempotent, OffsetDateTime approvedAt) {
        this.orderKey = orderKey;
        this.paymentKey = paymentKey;
        this.paymentState = paymentState;
        this.orderState = orderState;
        this.idempotent = idempotent;
        this.approvedAt = approvedAt;
    }

    public static PaymentApprovalResult approved(String orderKey, String paymentKey, OffsetDateTime approvedAt) {
        return new PaymentApprovalResult(
                orderKey,
                paymentKey,
                PaymentState.SUCCESS,
                OrderState.PAID,
                false,
                approvedAt
        );
    }

    public static PaymentApprovalResult alreadyApproved(String orderKey, String paymentKey, OffsetDateTime approvedAt) {
        return new PaymentApprovalResult(
                orderKey,
                paymentKey,
                PaymentState.SUCCESS,
                OrderState.PAID,
                true,
                approvedAt
        );
    }

    public static PaymentApprovalResult fromExisting(Payment payment) {
        OrderState orderState = (payment.getState() == PaymentState.SUCCESS) ? OrderState.PAID : OrderState.CREATED;
        return new PaymentApprovalResult(
                payment.getOrderKey(),
                payment.getExternalPaymentKey(),
                payment.getState(),
                orderState,
                true,
                payment.getPaidAt()
        );
    }

    public static PaymentApprovalResult failed(String orderKey, String paymentKey, OffsetDateTime approvedAt) {
        return new PaymentApprovalResult(
                orderKey,
                paymentKey,
                PaymentState.FAILED,
                OrderState.PAID,
                false,
                approvedAt
        );
    }

    public PaymentState getPaymentState() {
        return paymentState;
    }
}

File: core/core-domain/src/main/kotlin/com/coffee/gu/payment/PaymentApprovalResult.kt
package com.coffee.gu.payment

import com.coffee.gu.enums.OrderState
import com.coffee.gu.enums.PaymentState
import java.time.OffsetDateTime

class PaymentApprovalResult(
    val orderKey: String,
    val paymentKey: String,
    val paymentState: PaymentState,
    val orderState: OrderState,
    val idempotent: Boolean,
    val approvedAt: OffsetDateTime?,
) {
    companion object {
        fun approved(
            orderKey: String,
            paymentKey: String,
            approvedAt: OffsetDateTime,
        ): PaymentApprovalResult = PaymentApprovalResult(
            orderKey = orderKey,
            paymentKey = paymentKey,
            paymentState = PaymentState.SUCCESS,
            orderState = OrderState.PAID,
            idempotent = false,
            approvedAt = approvedAt,
        )

        fun alreadyApproved(
            orderKey: String,
            paymentKey: String,
            approvedAt: OffsetDateTime,
        ): PaymentApprovalResult = PaymentApprovalResult(
            orderKey = orderKey,
            paymentKey = paymentKey,
            paymentState = PaymentState.SUCCESS,
            orderState = OrderState.PAID,
            idempotent = true,
            approvedAt = approvedAt,
        )

        fun fromExisting(payment: Payment): PaymentApprovalResult {
            val orderState = if (payment.state == PaymentState.SUCCESS) OrderState.PAID else OrderState.CREATED
            return PaymentApprovalResult(
                orderKey = payment.orderKey,
                paymentKey = payment.externalPaymentKey ?: "",
                paymentState = payment.state,
                orderState = orderState,
                idempotent = true,
                approvedAt = payment.paidAt,
            )
        }

        fun failed(
            orderKey: String,
            paymentKey: String,
            approvedAt: OffsetDateTime,
        ): PaymentApprovalResult = PaymentApprovalResult(
            orderKey = orderKey,
            paymentKey = paymentKey,
            paymentState = PaymentState.FAILED,
            orderState = OrderState.PAID,
            idempotent = false,
            approvedAt = approvedAt,
        )
    }
}
